Allow non-admins to use `/engines/{model}/chat/completions`

msabramo opened the pull request:
The problem we were having is non-admin users trying to use `/engines/{model}/chat/completions` were getting an HTTP 401 error.

```shell
$ curl -sSL 'http://localhost:4000/engines/gpt-35-turbo-0125/chat/completions' \
    --header "Authorization: Bearer ${LITELLM_KEY}" \
    --header 'Content-Type: application/json' \
    --data ' {
        "model": "gpt-35-turbo-0125",
        "messages": [
            {
                "role": "user",
                "content": "Write a poem about LiteLLM"
            }
        ]
    }' \
    | jq '.'
{
  "error": {
    "message": "Authentication Error, Only proxy admin can be used to generate, delete, update info for new keys/users/teams. Route=/engines/gpt-35-turbo-0125/chat/completions. Your role=unknown. Your user_id=someone@company.com",
    "type": "auth_error",
    "param": "None",
    "code": 401
  }
}
```

This seems to be related to code in `user_api_key_auth` that checks that the URL matches a list of routes that are allowed for non-admin users, where the list of routes is in `LiteLLMRoutes.openai_routes.value`. The problem is that the route `/engines/{model}/chat/completions` is not in that list and furthermore, that wouldn't even work if it were, because the comparison is done with `request.url.path` and that will have the actual model name in it (e.g.: `gpt-35-turbo-0125`), rather than `{model}`.

I added a new list `LiteLLMRoutes.openai_route_names` and added the route **names** to that list. Then I added a check in `user_api_key_auth` to see if the route name is in the list of route names.

Aside, should this error be an HTTP 403 rather than an HTTP 401?

![Screenshot 2024-05-15 at 5 52 20 PM](https://github.com/BerriAI/litellm/assets/305268/295a712c-a33b-4e3d-9fc6-9295dc292466)

Cc: @taralika, @braintrain

msabramo committed:
Allow non-admins to use OpenAI routes

The problem we were having is non-admin users trying to use
`/engines/{model}/chat/completions` were getting an HTTP 401 error.

```shell
$ curl -sSL 'http://localhost:4000/engines/gpt-35-turbo-0125/chat/completions' \
    --header "Authorization: Bearer ${LITELLM_KEY}" \
    --header 'Content-Type: application/json' \
    --data ' {
        "model": "gpt-35-turbo-0125",
        "messages": [
            {
                "role": "user",
                "content": "Write a poem about LiteLLM"
            }
        ]
    }' \
    | jq '.'
{
  "error": {
    "message": "Authentication Error, Only proxy admin can be used to generate, delete, update info for new keys/users/teams. Route=/engines/gpt-35-turbo-0125/chat/completions. Your role=unknown. Your user_id=someone@company.com",
    "type": "auth_error",
    "param": "None",
    "code": 401
  }
}
```

This seems to be related to code in `user_api_key_auth` that checks that the URL
matches a list of routes that are allowed for non-admin users, where the list of
routes is in `LiteLLMRoutes.openai_routes.value`. The problem is that the route
`/engines/{model}/chat/completions` is not in that list and furthermore, that
wouldn't even work if it were, because the comparison is done with
`request.url.path` and that will have the actual model name in it (e.g.:
`gpt-35-turbo-0125`), rather than `{model}`.

I added a new list `LiteLLMRoutes.openai_route_names` and added the route
**names** to that list. Then I added a check in `user_api_key_auth` to see if
the route name is in the list of route names.

krrishdholakia commented on litellm/proxy/_types.py:
```diff
@@ -77,6 +77,16 @@ class LiteLLM_UpperboundKeyGenerateParams(LiteLLMBase):
+        "image_generation",
+        "audio_transcriptions",
+        "moderations",
+        "model_info_v1",
```
do you mean the `/v1/models` route - which is the openai-compatible endpoint 

or the litellm-specific `/model/info` route which is used for model management? 

given the name here, i'd assume you mean the `/v1/models` route? @msabramo

msabramo replied:
Thanks! Yes, I got a little confused. Updated the routes in b079f4cb7.

krrishdholakia commented on litellm/proxy/proxy_server.py:
```diff
@@ -1073,6 +1073,8 @@ async def user_api_key_auth(
             if not _is_user_proxy_admin(user_id_information):  # if non-admin
                 if route in LiteLLMRoutes.openai_routes.value:
                     pass
+                elif request['route'].name in LiteLLMRoutes.openai_route_names.value:
```
can you add a test for this? 

this might be a good place for it - https://github.com/BerriAI/litellm/blob/d208dedb35b2cec4ade6a2acb63f46eb84b49b3c/litellm/tests/test_key_generate_prisma.py#L4

msabramo replied:
I think I'm confused about how this relates to prisma and key generation, maybe because I haven't had enough coffee yet.

Can you give me an example or outline to bootstrap me?

krrishdholakia replied:
- create a virtual key (non-admin) 
- call user_api_key_auth with it, with the route you made this PR for 
- check if user_api_key_auth let's the request go through 

e.g. - https://github.com/BerriAI/litellm/blob/d208dedb35b2cec4ade6a2acb63f46eb84b49b3c/litellm/tests/test_key_generate_prisma.py#L149

msabramo replied:
How does d5b2e8e7e look?

krrishdholakia commented:
Hey @msabramo acknowledging the issue - added some comments

msabramo committed:
model_info_v{1,2} (LiteLLM) => model_list (OpenAI)

msabramo committed:
Make test_generate_and_call_with_valid_key parametrized

This allows us to test the same code with different routes.
For example, it lets us test the `/engines/{model}/chat/completions`
route, which https://github.com/BerriAI/litellm/pull/3663 fixes.

msabramo commented:
@krrishdholakia:

- Updated the routes in b079f4cb7.
- Added a test in d5b2e8e7e.

msabramo committed:
Add "/engines/{model}/chat/completions" to openai_routes

I don't think that this helps with the issue that I'm seeing, but I
think it might be nice to have this model listed in the openai_routes
list so that it's documented that it's a valid chat_completion route.

krrishdholakia commented:
Looks good. Can you share a screenshot of it passing `test_key_generate_prisma.py`? Should be good to merge after that. 

```
cd litellm/litellm
pytest test_key_generate_prisma.py -x -vv
```

krrishdholakia commented on litellm/proxy/_types.py:
```diff
@@ -77,6 +77,15 @@ class LiteLLM_UpperboundKeyGenerateParams(LiteLLMBase):
+        "image_generation",
+        "audio_transcriptions",
+        "moderations",
+        "model_list",  # OpenAI /v1/models route
```
i believe your test will fail for `/v1/models` since the endpoint name is `models` not `model_list`.

Can you sanity check this for me? @msabramo

msabramo replied:
Isn't it `model_list`?

```python
#### API ENDPOINTS ####
@router.get(
    "/v1/models", dependencies=[Depends(user_api_key_auth)], tags=["model management"]
)
@router.get(
    "/models", dependencies=[Depends(user_api_key_auth)], tags=["model management"]
)  # if project requires model list
def model_list(
    user_api_key_dict: UserAPIKeyAuth = Depends(user_api_key_auth),
):
```

krrishdholakia replied:
the function is - maybe i misread the code, i thought the name checked the url 

that's on me - no worries then. Thanks for clarifying.

msabramo replied:
No problem. I'm about to add a few more routes to the test.

msabramo replied:
cf7185735

msabramo committed:
Add more routes to test_generate_and_call_with_valid_key

msabramo committed:
Add more routes to test_generate_and_call_with_valid_key

msabramo committed:
Add nicer test ids when using pytest -v

Replace:

```
test_key_generate_prisma.py::test_generate_and_call_with_valid_key[api_route0] PASSED
test_key_generate_prisma.py::test_generate_and_call_with_valid_key[api_route10] PASSED
test_key_generate_prisma.py::test_generate_and_call_with_valid_key[api_route11] PASSED
test_key_generate_prisma.py::test_generate_and_call_with_valid_key[api_route12] PASSED
test_key_generate_prisma.py::test_generate_and_call_with_valid_key[api_route13] PASSED
test_key_generate_prisma.py::test_generate_and_call_with_valid_key[api_route14] PASSED
````

with:

```
litellm/tests/test_key_generate_prisma.py::test_generate_and_call_with_valid_key[{'route': 'audio_transcriptions', 'path': '/audio/transcriptions'}] PASSED
litellm/tests/test_key_generate_prisma.py::test_generate_and_call_with_valid_key[{'route': 'audio_transcriptions', 'path': '/v1/audio/transcriptions'}] PASSED
litellm/tests/test_key_generate_prisma.py::test_generate_and_call_with_valid_key[{'route': 'chat_completion', 'path': '/chat/completions'}] PASSED
litellm/tests/test_key_generate_prisma.py::test_generate_and_call_with_valid_key[{'route': 'chat_completion', 'path': '/engines/{model}/chat/completions'}] PASSED
litellm/tests/test_key_generate_prisma.py::test_generate_and_call_with_valid_key[{'route': 'chat_completion', 'path': '/openai/deployments/{model}/chat/completions'}] PASSED
litellm/tests/test_key_generate_prisma.py::test_generate_and_call_with_valid_key[{'route': 'chat_completion', 'path': '/v1/chat/completions'}] PASSED
```

msabramo commented:
With 4194bafae, I made the verbose pytest output more informative. New screenshot:

![Screenshot 2024-05-16 at 11 42 07 AM](https://github.com/BerriAI/litellm/assets/305268/89884093-120a-4fb1-8f14-72f80e3ce9ad)

krrishdholakia commented:
LGTM! thanks for the screenshot

msabramo commented:
I guess a Docker image will be built with this in a few hours? Just curious, is it easy to force a Docker image build to happen?

The reason I ask is I am about to do a knowledge-sharing session with two colleagues and it would be cool if we could deploy this change as a part of that, as it will unblock some other work that they might be involved in.

ishaan-jaff commented:
I can put out a dev release @msabramo, it will take 20 mins

ishaan-jaff commented:
It's queued - look out for 1.37.12.dev1 on the releases @msabramo

ishaan-jaff commented:
@msabramo it's here: https://github.com/BerriAI/litellm/pkgs/container/litellm/217335173?tag=main-v1.37.12.dev1

msabramo commented:
Awesome, thanks @ishaan-jaff!
